File: pyfenstein3d/engine/ray.py
import math
from .vector2d import Vector2d
from .wall import Wall
from .item import Item
from .door import Door


class Ray():
    def __init__(self, angle: float):
        self.__dist_adjusted = 0
        self.__dist = 0
        self.__offset: 0
        self.__is_vertical: False
        self.__is_inverted: False
        self.__type_id: int = None
        self.__rel_ang = angle
        self.__vector2d_ang = Vector2d(1, 0)
        self.__vector2d = Vector2d(0, 0)
        self.__items = []
        self.__doors = []
        self.rot(angle)

    @property
    def ang(self):
        return self.__vector2d_ang.ang

    @property
    def dir_x(self):
        return self.__dir_x

    @property
    def dir_y(self):
        return self.__dir_y

    @property
    def delta_dist_x(self):
        return self.__delta_dist_x

    @property
    def delta_dist_y(self):
        return self.__delta_dist_y

    @property
    def type_id(self):
        return self.__type_id

    @property
    def offset(self):
        return self.__offset

    @property
    def is_vertical(self):
        return self.__is_vertical

    @property
    def is_inverted(self):
        return self.__is_inverted

    @property
    def dist_adjusted(self):
        return self.__dist_adjusted

    @property
    def dist(self):
        return self.__dist

    @property
    def collided_vector2d(self) -> Vector2d:
        return self.__vector2d

    @property
    def items(self):
        return tuple(self.__items)

    @property
    def doors(self):
        return tuple(self.__doors)

    def cast_wall(self, pos: Vector2d, grid):
        dist_x = self.get_dist_x(pos)
        dist_y = self.get_dist_y(pos)
        block_x = math.floor(pos.x)
        block_y = math.floor(pos.y)
        while 0 <= block_x < grid.max_x or 0 <= block_y < grid.max_y:
            if dist_y < dist_x:
                block_y += self.dir_y
                item = grid.get_block(block_x, block_y)
                if isinstance(item, Wall):
                    self.__vector2d = Vector2d(
                        math.cos(self.ang) * dist_y,
                        math.sin(self.ang) * dist_y
                    ) + pos
                    self.__dist_adjusted = abs(
                        math.cos(self.__rel_ang) * dist_y)
                    self.__dist = abs(dist_y)
                    self.__is_inverted = pos.y > item.y
                    self.__offset = self.__vector2d.x % 1
                    self.__is_vertical = False
                    if self.__is_inverted and isinstance(grid.get_block(block_x, block_y + 1), Door):
                        self.__type_id = 50
                    elif not self.__is_inverted and isinstance(grid.get_block(block_x, block_y - 1), Door):
                        self.__type_id = 50
                    else:
                        self.__type_id = item.type_id

                    return
                dist_y += self.delta_dist_y
            else:
                block_x += self.dir_x
                item = grid.get_block(block_x, block_y)
                if isinstance(item, Wall):
                    self.__vector2d = Vector2d(
                        math.cos(self.ang) * dist_x,
                        math.sin(self.ang) * dist_x
                    ) + pos
                    self.__dist_adjusted = abs(
                        math.cos(self.__rel_ang) * dist_x)
                    self.__dist = abs(dist_x)
                    self.__is_inverted = pos.x > item.x
                    self.__offset = self.__vector2d.y % 1
                    self.__is_vertical = True
                    if self.__is_inverted and isinstance(grid.get_block(block_x + 1, block_y), Door):
                        self.__type_id = 50
                    elif not self.__is_inverted and isinstance(grid.get_block(block_x - 1, block_y), Door):
                        self.__type_id = 50
                    else:
                        self.__type_id = item.type_id
                    return
                dist_x += self.delta_dist_x

    def cast_doors(self, pos: Vector2d, doors: []):
        self.__doors = []
        for door in doors:
            if door.is_vertical:
                is_inverted = math.pi * 0.5 <= self.ang <= math.pi * 1.5
                if not is_inverted and  door.x > self.__vector2d.x:
                    continue
                if is_inverted and door.x < self.__vector2d.x:
                    continue
                x = door.x - pos.x;
                y = math.tan(self.ang) * x;
                v = Vector2d(x, y);
                pos_y = pos.y + y;
                if pos_y < door.y or pos_y > door.y + 1:
                    continue
                dist_adjusted = math.sin(math.pi / 2 - self.__rel_ang) * v.mag;
                offset = (door.y - pos_y) % 1 #- Math.floor(door.y - pos_y);
                self.__doors.append(RayDoor(dist_adjusted, offset, door.is_vertical))
                if v.mag < self.__dist:
                    self.__dist = v.mag
                    self.__vector2d = v + pos
                    self.__dist_adjusted = dist_adjusted
            else:
                is_inverted = math.pi <= self.ang <= math.pi * 2
                if not is_inverted and door.y > self.__vector2d.y:
                    continue
                if is_inverted and door.y < self.__vector2d.y:
                    continue
                y = door.y - pos.y
                x = y / math.tan(self.ang)
                v = Vector2d(x, y)
                pos_x = pos.x + x
                if pos_x < door.x or pos_x > door.x + 1:
                    continue
                dist_adjusted = math.sin(math.pi / 2 - self.__rel_ang) * v.mag
                offset = (door.x - pos_x) % 1
                self.__doors.append(RayDoor(dist_adjusted, offset, door.is_vertical))
                if v.mag < self.__dist:
                    self.__dist = v.mag
                    self.__vector2d = v + pos
                    self.__dist_adjusted = dist_adjusted

    def cast_items(self, player, items: []):
        self.__items = []
        pos = Vector2d(player.x, player.y)
        if self.type_id is not None:
            for i in items:
                item: Item = i
                item_pos = (Vector2d(item.x, item.y) - pos)
                ray_pos = (self.__vector2d - pos)
                if item_pos.mag < ray_pos.mag:
                    ray_pos = ray_pos ** -self.ang
                    item_pos = item_pos ** -self.ang
                    if ray_pos.x < item_pos.x:
                        continue
                    if item_pos.y > 0.5 or item_pos.y < -0.5:
                        continue
                    dist = math.sin(math.pi / 2 - self.__rel_ang) * item_pos.x
                    self.__items.append(RayItem(item.type_id, dist, item_pos.y, item.get_state(player)))
                    
                

    def rot(self, rad):
        self.__vector2d_ang = self.__vector2d_ang ** rad

        if self.ang > math.pi * 1.5 or self.ang < math.pi * 0.5:
            self.__dir_x = 1
        elif self.ang > math.pi * 0.5 and self.ang < math.pi * 1.5:
            self.__dir_x = -1
        else:
            self.__dir_x = 0

        if self.ang > 0 and self.ang < math.pi:
            self.__dir_y = 1
        elif self.ang > math.pi and self.ang < math.pi * 2:
            self.__dir_y = -1
        else:
            self.__dir_y = 0

        self.__delta_dist_x = abs(
            1 / math.cos(self.ang)) if math.cos(self.ang) != 0 else math.inf
        self.__delta_dist_y = abs(
            1 / math.sin(self.ang)) if math.sin(self.ang) != 0 else math.inf

    def get_dist_x(self, pos: Vector2d):
        if self.dir_x == 1:
            return abs((1 - pos.x % 1) / math.cos(self.ang))
        if self.dir_x == -1:
            return abs((pos.x % 1) / math.cos(self.ang))
        return math.inf

    def get_dist_y(self, pos: Vector2d):
        if self.dir_y == 1:
            return abs((1 - pos.y % 1) / math.sin(self.ang))
        if self.dir_y == -1:
            return abs((pos.y % 1) / math.sin(self.ang))
        return math.inf


class RayItem():
    def __init__(self, type_id, dist, offset, state):
        self.__type_id = type_id
        self.__dist = dist
        self.__offset = offset
        self.__state = state

    @property
    def type_id(self):
        return self.__type_id

    @property
    def dist(self):
        return self.__dist

    @property
    def offset(self):
        return self.__offset

    @property
    def state(self):
        return self.__state

class RayDoor():
    def __init__(self, dist, offset, is_vertical):
        self.__type_id = 49
        self.__dist = dist
        self.__offset = offset
        self.__is_vertical = is_vertical

    @property
    def type_id(self):
        return self.__type_id

    @property
    def dist(self):
        return self.__dist

    @property
    def offset(self):
        return self.__offset

    @property
    def is_vertical(self):
        return self.__is_vertical

    @property
    def state(self):
        return 1 if self.is_vertical else 0

File: engine/ray.go
package engine

import "math"

// Ray represents a single cast ray in the field of view.
type Ray struct {
	relAng       float64
	vector2dAng  *Vector2d
	vector2d     *Vector2d
	distAdjusted float64
	dist         float64
	offset       float64
	isVertical   bool
	isInverted   bool
	typeID       *int
	dirX         int
	dirY         int
	deltaDistX   float64
	deltaDistY   float64
	items        []*RayItem
	doors        []*RayDoor
}

// NewRay creates a new Ray at the given relative angle.
func NewRay(angle float64) *Ray {
	r := &Ray{
		relAng:      angle,
		vector2dAng: NewVector2d(1, 0),
		vector2d:    NewVector2d(0, 0),
		items:       []*RayItem{},
		doors:       []*RayDoor{},
	}
	r.Rot(angle)
	return r
}

func (r *Ray) Ang() float64          { return r.vector2dAng.Ang() }
func (r *Ray) DirX() int             { return r.dirX }
func (r *Ray) DirY() int             { return r.dirY }
func (r *Ray) DeltaDistX() float64   { return r.deltaDistX }
func (r *Ray) DeltaDistY() float64   { return r.deltaDistY }
func (r *Ray) TypeID() *int          { return r.typeID }
func (r *Ray) Offset() float64       { return r.offset }
func (r *Ray) IsVertical() bool      { return r.isVertical }
func (r *Ray) IsInverted() bool      { return r.isInverted }
func (r *Ray) DistAdjusted() float64 { return r.distAdjusted }
func (r *Ray) Dist() float64         { return r.dist }
func (r *Ray) CollidedVector2d() *Vector2d { return r.vector2d }

func (r *Ray) Items() []*RayItem { return r.items }
func (r *Ray) Doors() []*RayDoor { return r.doors }

// GetDistX returns the distance to the next x-boundary from pos.
func (r *Ray) GetDistX(pos *Vector2d) float64 {
	cosAng := math.Cos(r.Ang())
	if cosAng == 0 {
		return math.Inf(1)
	}
	if r.dirX == 1 {
		return math.Abs((1 - math.Mod(pos.X(), 1)) / cosAng)
	}
	if r.dirX == -1 {
		return math.Abs(math.Mod(pos.X(), 1) / cosAng)
	}
	return math.Inf(1)
}

// GetDistY returns the distance to the next y-boundary from pos.
func (r *Ray) GetDistY(pos *Vector2d) float64 {
	sinAng := math.Sin(r.Ang())
	if sinAng == 0 {
		return math.Inf(1)
	}
	if r.dirY == 1 {
		return math.Abs((1 - math.Mod(pos.Y(), 1)) / sinAng)
	}
	if r.dirY == -1 {
		return math.Abs(math.Mod(pos.Y(), 1) / sinAng)
	}
	return math.Inf(1)
}

// CastWall performs DDA ray-wall intersection against the grid.
func (r *Ray) CastWall(pos *Vector2d, grid *ItemGrid) {
	distX := r.GetDistX(pos)
	distY := r.GetDistY(pos)
	blockX := int(math.Floor(pos.X()))
	blockY := int(math.Floor(pos.Y()))

	for blockX < grid.MaxX() || blockY < grid.MaxY() {
		if distY < distX {
			blockY += r.dirY
			item := grid.GetBlock(blockX, blockY)
			if _, ok := item.(*Wall); ok {
				r.vector2d = NewVector2d(
					math.Cos(r.Ang())*distY,
					math.Sin(r.Ang())*distY,
				).Add(pos)
				r.distAdjusted = math.Abs(math.Cos(r.relAng) * distY)
				r.dist = math.Abs(distY)
				r.isInverted = pos.Y() > item.Y()
				r.offset = math.Mod(r.vector2d.X(), 1)
				r.isVertical = false
				belowOrAbove := blockY + 1
				if r.isInverted {
					belowOrAbove = blockY + 1
				} else {
					belowOrAbove = blockY - 1
				}
				if _, ok2 := grid.GetBlock(blockX, belowOrAbove).(*Door); ok2 {
					tid := 50
					r.typeID = &tid
				} else {
					tid := item.TypeID()
					r.typeID = &tid
				}
				return
			}
			distY += r.deltaDistY
		} else {
			blockX += r.dirX
			item := grid.GetBlock(blockX, blockY)
			if _, ok := item.(*Wall); ok {
				r.vector2d = NewVector2d(
					math.Cos(r.Ang())*distX,
					math.Sin(r.Ang())*distX,
				).Add(pos)
				r.distAdjusted = math.Abs(math.Cos(r.relAng) * distX)
				r.dist = math.Abs(distX)
				r.isInverted = pos.X() > item.X()
				r.offset = math.Mod(r.vector2d.Y(), 1)
				r.isVertical = true
				sideOrOther := blockX + 1
				if r.isInverted {
					sideOrOther = blockX + 1
				} else {
					sideOrOther = blockX - 1
				}
				if _, ok2 := grid.GetBlock(sideOrOther, blockY).(*Door); ok2 {
					tid := 50
					r.typeID = &tid
				} else {
					tid := item.TypeID()
					r.typeID = &tid
				}
				return
			}
			distX += r.deltaDistX
		}
	}
}

// CastDoors checks door intersections for visible doors.
func (r *Ray) CastDoors(pos *Vector2d, doors []*Door) {
	r.doors = []*RayDoor{}
	for _, door := range doors {
		if door.IsVertical() {
			isInverted := r.Ang() >= math.Pi*0.5 && r.Ang() <= math.Pi*1.5
			if !isInverted && door.X() > r.vector2d.X() {
				continue
			}
			if isInverted && door.X() < r.vector2d.X() {
				continue
			}
			x := door.X() - pos.X()
			y := math.Tan(r.Ang()) * x
			v := NewVector2d(x, y)
			posY := pos.Y() + y
			if posY < door.Y() || posY > door.Y()+1 {
				continue
			}
			distAdjusted := math.Sin(math.Pi/2-r.relAng) * v.Mag()
			offset := math.Mod(door.Y()-posY, 1)
			r.doors = append(r.doors, newRayDoor(distAdjusted, offset, door.IsVertical()))
			if v.Mag() < r.dist {
				r.dist = v.Mag()
				r.vector2d = v.Add(pos)
				r.distAdjusted = distAdjusted
			}
		} else {
			isInverted := r.Ang() >= math.Pi && r.Ang() <= math.Pi*2
			if !isInverted && door.Y() > r.vector2d.Y() {
				continue
			}
			if isInverted && door.Y() < r.vector2d.Y() {
				continue
			}
			y := door.Y() - pos.Y()
			x := y / math.Tan(r.Ang())
			v := NewVector2d(x, y)
			posX := pos.X() + x
			if posX < door.X() || posX > door.X()+1 {
				continue
			}
			distAdjusted := math.Sin(math.Pi/2-r.relAng) * v.Mag()
			offset := math.Mod(door.X()-posX, 1)
			r.doors = append(r.doors, newRayDoor(distAdjusted, offset, door.IsVertical()))
			if v.Mag() < r.dist {
				r.dist = v.Mag()
				r.vector2d = v.Add(pos)
				r.distAdjusted = distAdjusted
			}
		}
	}
}

// CastItems checks item/enemy intersections behind the wall hit.
func (r *Ray) CastItems(player *Player, items []Blocker) {
	r.items = []*RayItem{}
	pos := NewVector2d(player.X(), player.Y())
	if r.typeID != nil {
		for _, i := range items {
			itemPos := NewVector2d(i.X(), i.Y()).Sub(pos)
			rayPos := r.vector2d.Sub(pos)
			if itemPos.Mag() < rayPos.Mag() {
				rayPos = rayPos.Rot(-r.Ang())
				itemPos = itemPos.Rot(-r.Ang())
				if rayPos.X() < itemPos.X() {
					continue
				}
				if itemPos.Y() > 0.5 || itemPos.Y() < -0.5 {
					continue
				}
				dist := math.Sin(math.Pi/2-r.relAng) * itemPos.X()
				r.items = append(r.items, newRayItem(i.TypeID(), dist, itemPos.Y(), i.GetState(player)))
			}
		}
	}
}

// Rot rotates the ray by rad radians.
func (r *Ray) Rot(rad float64) {
	r.vector2dAng = r.vector2dAng.Rot(rad)

	ang := r.Ang()
	if ang > math.Pi*1.5 || ang < math.Pi*0.5 {
		r.dirX = 1
	} else if ang > math.Pi*0.5 && ang < math.Pi*1.5 {
		r.dirX = -1
	} else {
		r.dirX = 0
	}

	if ang > 0 && ang < math.Pi {
		r.dirY = 1
	} else if ang > math.Pi && ang < math.Pi*2 {
		r.dirY = -1
	} else {
		r.dirY = 0
	}

	cosAng := math.Cos(ang)
	if cosAng != 0 {
		r.deltaDistX = math.Abs(1 / cosAng)
	} else {
		r.deltaDistX = math.Inf(1)
	}
	sinAng := math.Sin(ang)
	if sinAng != 0 {
		r.deltaDistY = math.Abs(1 / sinAng)
	} else {
		r.deltaDistY = math.Inf(1)
	}
}

// RayItem stores information about an item seen by the ray.
type RayItem struct {
	typeID int
	dist   float64
	offset float64
	state  int
}

func newRayItem(typeID int, dist, offset float64, state int) *RayItem {
	return &RayItem{typeID: typeID, dist: dist, offset: offset, state: state}
}

func (ri *RayItem) TypeID() int    { return ri.typeID }
func (ri *RayItem) Dist() float64  { return ri.dist }
func (ri *RayItem) Offset() float64 { return ri.offset }
func (ri *RayItem) State() int     { return ri.state }

// RayDoor stores information about a door seen by the ray.
type RayDoor struct {
	typeID     int
	dist       float64
	offset     float64
	isVertical bool
}

func newRayDoor(dist, offset float64, isVertical bool) *RayDoor {
	return &RayDoor{typeID: 49, dist: dist, offset: offset, isVertical: isVertical}
}

func (rd *RayDoor) TypeID() int     { return rd.typeID }
func (rd *RayDoor) Dist() float64   { return rd.dist }
func (rd *RayDoor) Offset() float64 { return rd.offset }
func (rd *RayDoor) IsVertical() bool { return rd.isVertical }
func (rd *RayDoor) State() int {
	if rd.isVertical {
		return 1
	}
	return 0
}
